src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,514 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Generic
+{
+    /// <summary>
+    /// Represents a data structure in which each element additionally has a priority
+    /// associated with it.
+    /// </summary>
+    /// <typeparam name="TElement">The type of the element.</typeparam>
+    /// <typeparam name="TPriority">The type of the priority.</typeparam>
+    public class PriorityQueue<TElement, TPriority>
+    {
+        /// <summary>
+        /// Represents an implicit heap-ordered complete d-ary tree, stored as an array.
+        /// </summary>
+        private readonly List<(TElement element, TPriority priority)> nodes;
+
+        private const int RootIndex = 0;
+
+        /// <summary>
+        /// Specifies the arity of the d-ary heap, which here is quaternary.
+        /// </summary>
+        private const int Arity = 4;
+
+        /// <summary>
+        /// Creates an empty priority queue.
+        /// </summary>
+        public PriorityQueue()
+        {
+            this.nodes = new List<(TElement, TPriority)>();
+            this.UnorderedItems = new UnorderedItemsCollection(this.nodes);
+            this.Comparer = Comparer<TPriority>.Default;
```

Comment:
@eiriktsarpalis, the use a separate `IComparer<TPriority>` here rather than requiring `TPriority : IComparable<TPriority>` (or even just requiring `TElement : IComparable<TElement>` rather than having a separate priority) could have an impact on performance, e.g. for the latter if the TPriority is a struct, the JIT can likely devirtualize the interface calls, but for the former, every Compare operation is likely to be an interface dispatch.  Maybe that'll get a bit better when guarded devirtualization arrives, but it'd be interesting to understand before we commit this design exactly how much we're leaving on the table.

Reply:
Does the JIT optimize `Comparer<TPriority>.Default.Compare()` calls and is this something we could use to our advantage? (e.g. use separate sift implementations depending on whether the user has supplied a custom comparer).

Reply:
> Does the JIT optimize Comparer<TPriority>.Default.Compare() calls and is this something we could use to our advantage?

Currently, `Comparer<T>.Default.Compare` will be better than `_comparer.Compare` but not as good as `_value1.CompareTo(_value2)`.

Reply:
Requiring `TPriority : IComparable<TPriority>` would probably be fine for most uses, but I don't particularly like adopting `TElement : IComparable<TElement>` since it would result in more boilerplate for the element types.

> Currently, Comparer<T>.Default.Compare(_value1, _value2) will be [...] not as good as _value1.CompareTo(_value2).

Interesting, what is the main difference and what types does it concern? When I checked for numeric types it seemed to have applied all relevant optimizations.

Reply:
> Interesting, what is the main difference and what types does it concern? When I checked for numeric types it seemed to have applied all relevant optimizations.

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Diagnosers;
using BenchmarkDotNet.Running;
using System;
using System.Collections.Generic;

public class Program
{
    static void Main(string[] args) => BenchmarkSwitcher.FromAssemblies(new[] { typeof(Program).Assembly }).Run(args);
}

[DisassemblyDiagnoser(maxDepth: 3)]
[GenericTypeArguments(typeof(int))]
public class Tests<T> where T : IComparable<T>
{
    private IComparer<T> _comparer = Comparer<T>.Default;
    private T _a, _b;

    [Benchmark]
    public int ComparerInterface() => _comparer.Compare(_a, _b);

    [Benchmark]
    public int ComparerDirect() => Comparer<T>.Default.Compare(_a, _b);

    [Benchmark]
    public int CompareTo() => _a.CompareTo(_b);
}
```

|            Method |      Mean |
|------------------ |----------:|
| ComparerInterface | 1.8169 ns |
|    ComparerDirect | 1.3760 ns |
|         CompareTo | 0.5979 ns |

## .NET Core 5.0 (CoreCLR 42.42.42.42424, CoreFX 42.42.42.42424), X64 RyuJIT
```assembly
; Tests`1[[System.Int32, System.Private.CoreLib]].ComparerInterface()
       mov       [rsp+8],rcx
       mov       rcx,[rcx+8]
       mov       rdx,[rsp+8]
       mov       r8d,[rdx+14]
       mov       edx,[rdx+10]
       mov       r11,7FFD0F800638
       mov       rax,7FFD0F800638
       mov       rax,[rax]
       jmp       rax
; Total bytes of code 47
```

## .NET Core 5.0 (CoreCLR 42.42.42.42424, CoreFX 42.42.42.42424), X64 RyuJIT
```assembly
; Tests`1[[System.Int32, System.Private.CoreLib]].ComparerDirect()
       mov       edx,[rcx+10]
       mov       r8d,[rcx+14]
       mov       rcx,21332F32CD0
       mov       rcx,[rcx]
       mov       rax,offset System.Collections.Generic.GenericComparer`1[[System.Int32, System.Private.CoreLib]].Compare(Int32, Int32)
       mov       rax,[rax]
       jmp       rax
; Total bytes of code 36
```
```assembly
; System.Collections.Generic.GenericComparer`1[[System.Int32, System.Private.CoreLib]].Compare(Int32, Int32)
       cmp       edx,r8d
       jge       short M01_L00
       mov       eax,0FFFFFFFF
       jmp       short M01_L02
M01_L00:
       cmp       edx,r8d
       jle       short M01_L01
       mov       eax,1
       jmp       short M01_L02
M01_L01:
       xor       eax,eax
M01_L02:
       ret
; Total bytes of code 27
```

## .NET Core 5.0 (CoreCLR 42.42.42.42424, CoreFX 42.42.42.42424), X64 RyuJIT
```assembly
; Tests`1[[System.Int32, System.Private.CoreLib]].CompareTo()
       lea       rax,[rcx+10]
       mov       edx,[rcx+14]
       cmp       [rax],edx
       jge       short M00_L00
       mov       eax,0FFFFFFFF
       jmp       short M00_L02
M00_L00:
       cmp       [rax],edx
       jle       short M00_L01
       mov       eax,1
       jmp       short M00_L02
M00_L01:
       xor       eax,eax
M00_L02:
       ret
; Total bytes of code 32
```

Reply:
> I was looking at it via sharplab which was showing a different picture

FWIW, SharpLab does show a similar picture:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8lAbhvoYGUALbKAA4AZbMAB0AJQCuAOwwBLfDHbUaAbQBS8jAHEYMmFHlgAFBgCeAmBABmJ+XICUjgLocAzE1IMAwgB4AFQA+BgB3HkMYBgCGEAYASR8IfAF+UQAbGECgmgBvGgZChgEjADdsDCjE5NTYKGyGAH1IFP5DBgBeXxq2+uCxABEYG2wpdIwVIuKyiqiYxuw0JuAVAqLiTwcMbta6+LlDEbAYE0dOkOaeurEk3ZOFpcbgR1XqKY2GLZ3awwH5WDAGFO52+vWyg2Go3GNyu90WyxeNDWhQ+X1uPwCEGBHQu2Bhd0xJieiOoAF8gA=
You just need to make sure to select x64 as the target.  And it doesn't show the targets of the jmps for the interface/direct cases.

> Perhaps using Comparer.Default directly is a good enough compromise.

Perhaps.  It would likely mean multiple implementations of the key methods, chosen based on an if check for whether the _comparer was null or not (and not initializing it in the ctor), or potentially with an object equality check for whether it's the default.  I think it'd still be worth measuring against an alternate prototype that constrained TPriority, just to understand if there's really a "there" there or if this is just academic and I should be quiet ;-)

> Should we decide to change to TPriority : IComparable<TPriority>, would we need to take the changes through API review again?

Probably, just as a sanity check, in particular because it would also mean removing anything to do with IComparer.

Reply:
One downside of moving to `IComparable<T>` is that we would now have to add null checks to priorities.

Reply:
That's a maintenance downside rather than a perf downside: the comparer already needs to do such null checks.

Reply:
I ran a benchmark using an `IComparable<T>` prototype and the performance improvements are indeed impressive:

|                   Method |    Size |             Mean |            Error |           StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------- |-------- |-----------------:|-----------------:|-----------------:|------:|------:|------:|----------:|
|            PriorityQueue |      10 |         611.2 ns |          9.29 ns |          8.69 ns |     - |     - |     - |         - |
| PriorityQueue_Comparable |      10 |         399.0 ns |          5.51 ns |          5.15 ns |     - |     - |     - |         - |
|            PriorityQueue |      50 |       5,322.6 ns |         64.15 ns |         53.57 ns |     - |     - |     - |         - |
| PriorityQueue_Comparable |      50 |       2,566.8 ns |         29.87 ns |         27.94 ns |     - |     - |     - |         - |
|            PriorityQueue |     150 |      23,805.6 ns |        381.52 ns |        356.87 ns |     - |     - |     - |         - |
| PriorityQueue_Comparable |     150 |      11,975.8 ns |         81.13 ns |         67.75 ns |     - |     - |     - |         - |
|            PriorityQueue |     500 |     105,950.0 ns |      1,185.78 ns |      1,051.16 ns |     - |     - |     - |         - |
| PriorityQueue_Comparable |     500 |      67,355.3 ns |        350.80 ns |        328.14 ns |     - |     - |     - |         - |
|            PriorityQueue |    1000 |     234,004.0 ns |      1,756.55 ns |      1,557.13 ns |     - |     - |     - |         - |
| PriorityQueue_Comparable |    1000 |     154,620.7 ns |        430.27 ns |        359.30 ns |     - |     - |     - |         - |
|            PriorityQueue |   10000 |   3,137,616.1 ns |     38,805.91 ns |     36,299.07 ns |     - |     - |     - |         - |
| PriorityQueue_Comparable |   10000 |   1,969,164.0 ns |     24,864.32 ns |     23,258.10 ns |     - |     - |     - |         - |
|            PriorityQueue | 1000000 | 545,824,881.8 ns | 10,542,900.12 ns | 12,947,628.15 ns |     - |     - |     - |         - |
| PriorityQueue_Comparable | 1000000 | 374,415,631.2 ns |  7,471,951.52 ns | 11,632,928.33 ns |     - |     - |     - |         - |

Next step is to compare with an implementation that calls `Comparer<T>.Default.Compare()` directly in the default case.

Reply:
And here are the results when [an implementation calling Comparer<T>.Default.Compare() directly](https://github.com/eiriktsarpalis/pq-tests/commit/fe7e10c296a30ddd92087b51c5b304963e51993a) is added:

|                       Method |    Size |             Mean |           Error |           StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------------- |-------- |-----------------:|----------------:|-----------------:|------:|------:|------:|----------:|
|                PriorityQueue |      10 |         587.0 ns |         5.70 ns |          5.33 ns |     - |     - |     - |         - |
| PriorityQueue_InlineComparer |      10 |         504.1 ns |         5.85 ns |          4.89 ns |     - |     - |     - |         - |
|     PriorityQueue_Comparable |      10 |         336.9 ns |         6.65 ns |          6.23 ns |     - |     - |     - |         - |
|                PriorityQueue |      50 |       5,157.0 ns |        96.71 ns |        153.39 ns |     - |     - |     - |         - |
| PriorityQueue_InlineComparer |      50 |       4,120.3 ns |        62.49 ns |         55.39 ns |     - |     - |     - |         - |
|     PriorityQueue_Comparable |      50 |       2,576.7 ns |        49.96 ns |         46.73 ns |     - |     - |     - |         - |
|                PriorityQueue |     150 |      21,580.4 ns |       431.17 ns |        442.78 ns |     - |     - |     - |         - |
| PriorityQueue_InlineComparer |     150 |      18,798.7 ns |       127.74 ns |         99.73 ns |     - |     - |     - |         - |
|     PriorityQueue_Comparable |     150 |      12,086.9 ns |       175.85 ns |        164.49 ns |     - |     - |     - |         - |
|                PriorityQueue |     500 |     101,169.7 ns |     2,015.01 ns |      2,620.08 ns |     - |     - |     - |         - |
| PriorityQueue_InlineComparer |     500 |      91,078.9 ns |     1,634.16 ns |      1,528.60 ns |     - |     - |     - |         - |
|     PriorityQueue_Comparable |     500 |      66,140.8 ns |       902.69 ns |        753.79 ns |     - |     - |     - |         - |
|                PriorityQueue |    1000 |     218,271.4 ns |     3,696.23 ns |      3,457.46 ns |     - |     - |     - |         - |
| PriorityQueue_InlineComparer |    1000 |     190,335.2 ns |     2,882.19 ns |      2,696.00 ns |     - |     - |     - |         - |
|     PriorityQueue_Comparable |    1000 |     145,351.5 ns |       775.00 ns |        687.02 ns |     - |     - |     - |         - |
|                PriorityQueue |   10000 |   2,918,805.4 ns |    56,459.89 ns |     52,812.61 ns |     - |     - |     - |         - |
| PriorityQueue_InlineComparer |   10000 |   2,526,257.2 ns |    50,279.34 ns |     47,031.32 ns |     - |     - |     - |         - |
|     PriorityQueue_Comparable |   10000 |   1,677,831.3 ns |    22,245.65 ns |     20,808.60 ns |     - |     - |     - |         - |
|                PriorityQueue | 1000000 | 478,687,587.5 ns | 9,072,163.88 ns |  8,910,077.99 ns |     - |     - |     - |         - |
| PriorityQueue_InlineComparer | 1000000 | 447,056,846.7 ns | 7,063,243.07 ns |  6,606,961.95 ns |     - |     - |     - |    1336 B |
|     PriorityQueue_Comparable | 1000000 | 333,116,831.6 ns | 6,570,646.37 ns | 14,284,044.45 ns |     - |     - |     - |         - |

I think we should just change the signature and require `TPriority : IComparable`.

Reply:
@AndyAyersMS, your change in https://github.com/dotnet/coreclr/commit/ccc5e17738a4b2f652059f31e8146513163dbee7 made `EqualityComparer<T>.Default.Equals` fully inlineable.  Any thoughts on `Comparer<T>.Default.Compare`?

Reply:
I think we should just go ahead and change `TPriority : IComparable<TPriority>`. It significantly improves performance and simplifies the API signature. Only real downsides is not supporting `null` priorities and potential boilerplate when custom comparisons are required, but realistically neither seems likely to impact users. Should we just change it in this PR?

Reply:
Another disadvantage of `TPriority : IComparable<TPriority>` is being inconsistent with e.g `SortedSet<T>` and `SortedDictionary<TKey,TValue>`. What's the impact of

> potential boilerplate when custom comparisons are required

?

There is value in releasing the alpha version for testing by customers, in a way that is consistent with other data structures, i.e. with using `IComparer`.

Reply:
> What's the impact of
>
> > potential boilerplate when custom comparisons are required
>
> ?

I was thinking about users needing to define their own priority type with a custom `IComparable<TPriority>` implementation if the existing types don't satisfy their needs. You cannot change the ordering of, say, strings without wrapping it into a custom type. In practice though this can be easily worked around depending on the use case.

> There is value in releasing the alpha version for testing by customers, in a way that is consistent with other data structures, i.e. with using IComparer.

I would tend to agree in general, however the performance ramifications are too dramatic to ignore.

Reply:
The JIT team are working on optimizing `Comparer<T>.Default.CompareTo` calls, so the best course of action is to stick with the existing `IComparer` design.

Cf. https://github.com/dotnet/runtime/issues/39873